File: packages/core/lib/src/arabic.dart
/// Matching what a person typed against what a place is called.
///
/// A passenger looking for ملكا may type ملكة, مَلكا, or الملكا. She should find
/// it. Arabic search that only does exact string equality fails constantly —
/// on hamza, on ta marbuta, on the definite article, on the diacritics a
/// keyboard sometimes inserts — and each failure looks to her like the app not
/// knowing her village.
library;

const _diacritics = 'ًٌٍَُِّْٰٕٓٔ';
const _tatweel = 'ـ';

/// Folds the differences that are spelling rather than meaning.
String foldArabic(String input) {
  final buffer = StringBuffer();

  for (final rune in input.runes) {
    final ch = String.fromCharCode(rune);
    if (_diacritics.contains(ch) || ch == _tatweel) continue;

    buffer.write(switch (ch) {
      'أ' || 'إ' || 'آ' || 'ٱ' => 'ا',
      'ى' => 'ي',
      'ة' => 'ه',
      'ؤ' => 'و',
      'ئ' => 'ي',
      // Eastern Arabic digits, which a keyboard may produce mid-word.
      '٠' => '0', '١' => '1', '٢' => '2', '٣' => '3', '٤' => '4',
      '٥' => '5', '٦' => '6', '٧' => '7', '٨' => '8', '٩' => '9',
      _ => ch.toLowerCase(),
    });
  }

  var folded = buffer.toString().trim();
  // The definite article is optional in speech and in typing.
  if (folded.startsWith('ال') && folded.length > 3) folded = folded.substring(2);
  return folded.replaceAll(RegExp(r'\s+'), ' ');
}

/// Edit distance, capped so a long comparison stops early.
int _editDistance(String a, String b, {int max = 2}) {
  if ((a.length - b.length).abs() > max) return max + 1;

  var previous = List<int>.generate(b.length + 1, (i) => i);
  for (var i = 1; i <= a.length; i++) {
    final current = <int>[i, ...List<int>.filled(b.length, 0)];
    for (var j = 1; j <= b.length; j++) {
      final cost = a.codeUnitAt(i - 1) == b.codeUnitAt(j - 1) ? 0 : 1;
      current[j] = [current[j - 1] + 1, previous[j] + 1, previous[j - 1] + cost]
          .reduce((x, y) => x < y ? x : y);
    }
    previous = current;
  }
  return previous[b.length];
}

/// True when [query] plausibly names [candidate].
///
/// Folding handles the differences that are spelling. One further edit is
/// allowed on names long enough for it to be safe, which catches the ordinary
/// typo and the autocorrect that turns ملكا into ملكة — village names are far
/// enough apart that this does not confuse two of them.
bool matchesPlace(String query, String candidate) {
  final q = foldArabic(query);
  final c = foldArabic(candidate);
  if (q.isEmpty) return false;

  if (c.contains(q) || q.contains(c)) return true;
  if (q.length < 4 || c.length < 4) return false;
  return _editDistance(q, c, max: 1) <= 1;
}

File: apps/web-passenger/src/arabic.ts
/**
 * Matching what a person typed against what a place is called.
 *
 * A passenger looking for ملكا may type ملكة, مَلكا, or الملكا. She should find
 * it. Arabic search that only does exact string equality fails constantly — on
 * hamza, on ta marbuta, on the definite article, on the diacritics a keyboard
 * sometimes inserts — and each failure looks to her like the app not knowing
 * her village.
 *
 * A direct port of packages/core/lib/src/arabic.dart. The two are tested
 * against the same cases so the web and native clients agree on what a name is.
 */

const DIACRITICS = "ًٌٍَُِّْٰٕٓٔ";
const TATWEEL = "ـ";

const FOLD: Record<string, string> = {
  "أ": "ا", "إ": "ا", "آ": "ا", "ٱ": "ا",
  "ى": "ي",
  "ة": "ه",
  "ؤ": "و",
  "ئ": "ي",
  // Eastern Arabic digits, which a keyboard may produce mid-word.
  "٠": "0", "١": "1", "٢": "2", "٣": "3", "٤": "4",
  "٥": "5", "٦": "6", "٧": "7", "٨": "8", "٩": "9",
};

/** Folds the differences that are spelling rather than meaning. */
export function foldArabic(input: string): string {
  let out = "";
  for (const ch of input) {
    if (DIACRITICS.includes(ch) || ch === TATWEEL) continue;
    out += FOLD[ch] ?? ch.toLowerCase();
  }

  let folded = out.trim();
  // The definite article is optional in speech and in typing.
  if (folded.startsWith("ال") && folded.length > 3) folded = folded.slice(2);
  return folded.replace(/\s+/g, " ");
}

/** Edit distance, capped so a long comparison stops early. */
function editDistance(a: string, b: string, max = 2): number {
  if (Math.abs(a.length - b.length) > max) return max + 1;

  let previous = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 1; i <= a.length; i++) {
    const current = [i, ...new Array<number>(b.length).fill(0)];
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      current[j] = Math.min(current[j - 1] + 1, previous[j] + 1, previous[j - 1] + cost);
    }
    previous = current;
  }
  return previous[b.length];
}

/**
 * True when `query` plausibly names `candidate`.
 *
 * Folding handles the differences that are spelling. One further edit is
 * allowed on names long enough for it to be safe, which catches the ordinary
 * typo and the autocorrect that turns ملكا into ملكة — village names are far
 * enough apart that this does not confuse two of them.
 */
export function matchesPlace(query: string, candidate: string): boolean {
  const q = foldArabic(query);
  const c = foldArabic(candidate);
  if (!q) return false;

  if (c.includes(q) || q.includes(c)) return true;
  if (q.length < 4 || c.length < 4) return false;
  return editDistance(q, c, 1) <= 1;
}
